### ISSUE TITLE
Use Language Server Protocol (LSP) to re-implement all code editing

### ISSUE THREAD

**Issue by xingyaoww** [All-Hands-AI/OpenHands#1934]
**What problem or use case are you trying to solve?**

OpenDevin now sometimes suffers with edit command (concrete example: https://github.com/OpenDevin/OpenDevin/pull/1928).

**Describe the UX of the solution you'd like**

As suggested by a friend @ganler, maybe it is more like a fundamental solution if we can gradually replace all the `swe-agent` tools to interface with a [Language Server Protocol](https://microsoft.github.io/language-server-protocol/overviews/lsp/overview/):

![image](https://github.com/OpenDevin/OpenDevin/assets/38853559/1d1ab2e9-8ea5-4f74-8d0d-a3c02d7bcbe7)

It gives us the ability to:
1. Open file
2. Change/edit file
3. Go to the definition of a symbol
4. More importantly, this is actually [scale to multiple programming languages](https://microsoft.github.io/language-server-protocol/implementors/servers/)!

**Do you have thoughts on the technical implementation?**

- [ ] Starts with translating the existing bash command's implementation to an API call to the language server (e.g., open, edit, etc)
- [ ] Figure out a better interface for editing -- current HEREDOC + bash causes many weird issues. Can we potentially use `IPython` capability for editing, write these file operation functions in Python, and have the model write and execute Python cells to perform file operations? -- can directly wrap around https://github.com/microsoft/monitors4codegen?tab=readme-ov-file#4-multilspy
- [ ] Support more advanced features, like going to the definition
- [ ] Automatically supports multiple language servers for most existing programming languages.
- [ ] Show VSCode warning / problems

**Describe alternatives you've considered**

**Additional context**

**Comment by li-boxuan** [comment]
Oh nice! I have worked with LSP before! I could probably contribute to this.

**Comment by ganler** [comment]
LSP has a general interface to interact with. But if we want to start with some very easy, Jedi can be a great starter in my experience. :) 

https://jedi.readthedocs.io/en/latest/docs/api.html#examples

**Comment by ganler** [comment]
For more general cases of interactions, `monitors4codegen` has some examples using `multilspy` to interact with some LSP implementations.

https://github.com/microsoft/monitors4codegen#4-multilspy

**Comment by casper-hansen** [comment]
FYI, Devin (the closed one) is fully integrated with VS Code. That means it has access to all the extensions, linting, and debugging tools. I think this is something to consider

**Comment by frankxu2004** [comment]
I wonder if LSP can also actually edit the files, in addition to just monitoring file-changes or provide completion suggestions, etc.

**Comment by ramsey-coding** [comment]
Is it a good idea to incorporate complex LSP protocol with agent?

**Comment by ganler** [comment]
> I wonder if LSP can also actually edit the files, in addition to just monitoring file changes (usually user-triggered) and provide completion suggestions, etc.
> 
> For the VSCode solution, maybe incorporating the full VScode is heavy, but the underlying editor of vscode, monaco, could be a good option. Alternatively, CodeMirror is also quite minimalist. These editors usually maintain editor state, including undo stack as well so that if the agent is doing something wrong, we can let it call "undo" function.
> 
> For the `edit` command replacement, the closest might be: https://microsoft.github.io/monaco-editor/typedoc/interfaces/editor.ICodeEditor.html#executeEdits , or https://codemirror.net/examples/change/

LSP models cursors and code locations & ranges. And knowing the locations it's easy to insert code, isn't it ;).

**Comment by frankxu2004** [comment]
Good point. @ganler I wonder who maintains the cursor and code location states? (the editor states). For a real editor there's actual editor states, but it seems like we still have to maintain an "headless" editor in the backend

**Comment by 0xdevalias** [comment]
This may also be another alternative/useful exploration:

- #742

**Comment by xingyaoww** [comment]
cc @ryanhoangt who is potentially interested in this

**Comment by ryanhoangt** [comment]
Hi @LakshyAAAgrawal, thanks for offering your help! I've tried to integrate `multilspy` into the [openhands-aci](https://github.com/All-Hands-AI/openhands-aci/) before (at [this branch](https://github.com/All-Hands-AI/openhands-aci/tree/ht/integrate-lsp)) but didn't finish it. Would be great if you can have a look to see if we can integrate it as a module into the aci!

**Comment by LakshyAAAgrawal** [comment]
Hi all!

I apologize for not being able to follow up on this soon. Does anyone have ideas/suggestions for what are the main parts that require multilspy (https://aka.ms/multilspy) integration? My suggestion would be to start with one small feature that some current solution can be replaced by lsp, and then slowly expand the set of locations. Since I am new to Openhands, it would be great if anyone could advise about what might be the best place to start with integration!

**Comment by xingyaoww** [comment]
This is probably not planned for the team now. We currently implement every file editing in the https://github.com/All-Hands-AI/openhands-aci repo -- We can possibly implement ACI there.

Happy to re-open this if anyone in the community is interested in implementing this :)
